### PR TITLE
Digest authentication improvements

### PR DESCRIPTION
**Pull request by markstory**
Implement more robust nonce handling in Digest authentication. Shore up some of the implementation gaps we have around nonces. Nonces should expire and should be generated by the server. I've followed the Symfony approach to generating nonces which results in short lived (5 minutes by default) nonces that are regenerated each time the client receives a 401.

By validating and requiring nonce rotation we can avoid replay attacks with our Digest authentication implementation.

I've targetted 3.next as validating/requiring nonce rotation is a behavior change that I feel could trip up application developers. By including it in a minor release we can more effectively communicate the changes.

Refs #9668

### CODE REVIEW

File: src/Auth/DigestAuthenticate.php [hunk 1]
```diff
@@ -69,11 +71,12 @@ class DigestAuthenticate extends BasicAuthenticate
      * Besides the keys specified in BaseAuthenticate::$_defaultConfig,
      * DigestAuthenticate uses the following extra keys:
      *
+     * - `secret` The secret to use for nonce validation. Defaults to Security.salt.
      * - `realm` The realm authentication is for, Defaults to the servername.
-     * - `nonce` A nonce used for authentication. Defaults to `uniqid()`.
      * - `qop` Defaults to 'auth', no other values are supported at this time.
      * - `opaque` A string that must be returned unchanged by clients.
      *    Defaults to `md5($config['realm'])`
+     * - `nonceExpires` The number of seconds that nonces are valid for. Defaults to 300.
```
Comment (jippi) [inline src/Auth/DigestAuthenticate.php]: `nonceLifetime` ?

**Comment by lorenzo** [comment]
Nice job! @markstory

**Comment by JelmerD** [comment]
Awesome, I'll give it a go! And thank you very much for the quick pick-up!

**Comment by markstory** [comment]
@JelmerD Thanks for looking at the implementation so thoroughly.

**Pull request merged**